drafts/records.md

```diff
@@ -0,0 +1,398 @@
+A `record` may be used as a `readonly class`,
+as the behavior of it is very similar with no key differences to assist in migration from `readonly class`.
+
+#### Optional parameters and default values
+
+A `record` can also be defined with optional parameters that are set if omitted during instantiation.
+
+``` php
+record Rectangle(int $x, int $y = 10);
+var_dump(Rectangle(10)); // output a record with x: 10 and y: 10
+```
+
+#### Auto-generated `with` method
+
+To enhance the usability of records, the RFC proposes automatically generating a `with` method for each record.
+This method allows for partial updates of properties, creating a new instance of the record with the specified
+properties updated.
+
+The auto-generated `with` method accepts only named arguments defined in the constructor.
+No other property names can be used, and it returns a record object with the given values.
+
+``` php
+$point1 = Point(3, 4);
+$point2 = $point1->with(x: 5);
+$point3 = $point1->with(null, 10); // must use named arguments
+
+echo $point1->x; // Outputs: 3
+echo $point2->x; // Outputs: 5
+```
+
+A developer may define their own `with` method if they so choose,
+and reference the generated `with` method using `parent::with()`.
+This allows a developer to define policies or constraints on how data is updated.
+
+``` php
+record Planet(string $name, int $population) {
+  public function with(int $population) {
+    return parent::with(population: $population);
+  }
+}
+$pluto = Planet("Pluto", 0);
+// we made it!
+$pluto = $pluto->with(population: 1);
+// and then we changed the name
+$mickey = $pluto->with(name: "Mickey"); // no named argument for population error
+```
+
+#### Constructors
+
+Optionally, they may also define a constructor to provide validation or other initialization logic:
+
+```php
+record User(string $name, string $email) {
+  public string $id;
+
+  public function __construct() {
+    if (!filter_var($email, FILTER_VALIDATE_EMAIL)) {
+      throw new InvalidArgumentException("Invalid email address");
+    }
+    
+    $this->id = hash('sha256', $email);
+    $this->name = ucwords($name);
+  }
+}
+```
+
+During construction, a `record` is fully mutable.
+This allows the developer freedom to mutate properties as needed to ensure a canonical representation of an object.
+
+### Performance considerations
+
+To ensure that records are both performant and memory-efficient,
+the RFC proposes leveraging PHP's copy-on-write (COW) semantics (similar to arrays) and interning values.
+Unlike interned strings, the garbage collector will be allowed to clean up these interned records when they are no
+longer needed.
+
+``` php
+$point1 = Point(3, 4);
+$point2 = $point1; // No data duplication, $point2 references the same data as $point1
+$point3 = Point(3, 4); // No data duplication here either, it is pointing the the same memory as $point1
+
+$point4 = $point1->with(x: 5); // Data duplication occurs here, creating a new instance with modified data
```

Comment:
Shorten `data duplication` to `data dups`
Drop `the` and `the the`?

These ^^ to keep lines shorter and more readable _(and to avoid rendering with a horizontal scroll bar when the browser is not full screen)_